### DISCUSSION TITLE
how to use delta as pager/diff tool?

### DISCUSSION THREAD

**Discussion by chiefjester** [PatrickF1/fzf.fish#202] [Q&A]
Hi, 👋 

I've recently used [dandavison/delta/](https://github.com/dandavison/delta/) to get better diffs.

I was wondering if there's a way to use it in the git log preview? <kbd>ctrl</kbd>+<kbd>alt</kbd>+<kbd>L</kbd>

I see that in [fzf.vim](https://github.com/junegunn/fzf.vim#dependencies) it uses delta if it detects it's installed.

**Comment by notDavid** [comment]
Fyi for anyone else stumbling upon this and wondering why it isn’t working for them;

I’ve [set delta as my pager in gitconfig](https://github.com/dandavison/delta/#get-started), and something like `git show 12345` works and will show me delta’s output.
However for some reason, when using fzf.fish “git log”, the preview window (which shows the same `git show` output) does **not** show delta’s output. I wonder why?

Anyway, setting this environment variable (in the fish shell) works for me:
`set -x fzf_git_log_opts --preview='git show {1} | delta'`

**Reply by kidonng** [reply]
> However for some reason, when using fzf.fish “git log”, the preview window (which shows the same `git show` output) does **not** show delta’s output. I wonder why?

Because the output of `git show` is piped into `fzf`, so Git won't call the pager.

**Comment by bagohart** [comment]
The solution proposed by @notDavid  works for git log, but it doesn't work for git status.
After a bit of tinkering I made it work by setting the `$fzf_git_status_opts` variable:
`set -gx fzf_git_status_opts '--preview=fzf_preview_changed_file_using_delta {}`
and adding a new function `fzf_preview_changed_file_using_delta`:

```fish
function fzf_preview_changed_file_using_delta
    set -l path (string split ' ' $argv)[-1]
    if string match -r '^\?\?' $argv --quiet
        echo -e (set_color --underline)=== Untracked ===\n
        _fzf_preview_file $path
    else
        if string match -r '\S\s\S' $argv --quiet
            echo -e (set_color --underline red)=== Unstaged ===\n
            git diff --color=always -- $path | delta
        end
        if string match -r '^\S' $argv --quiet
            echo -e (set_color --underline green)=== Staged ===\n
            git diff --color=always --staged -- $path | delta
        end
    end
end
```

That function is copied and adapted from `_fzf_preview_changed_file` to pipe the output of `git diff` through `delta`.

That solution is not ideal though because I will now miss updates to `_fzf_preview_changed_file`.

**Comment by PatrickF1** [comment]
I'm currently looking into making this a feature out of the box based on ideas from https://github.com/dandavison/delta/discussions/840. Stay tuned! And check out that discussion for ideas for yourself in the meantime.

**Reply by PatrickF1** [reply]
Discussed it with the git contributors here https://lore.kernel.org/git/20230816025715.GB2248431@coredump.intra.peff.net/T/#t

**Reply by PatrickF1** [reply]
Ok based on the Jeff King (the git contributor who replied) feedback, the feature I will build will read the pager command from `interactive.diffFilter` and if it exists, pipe the output of git show through it. So if you want to integrate your diff tool with fzf.fish, you will need to add the appropriate pager to `interactive.diffFilter`. Appropriate meaning it is appropriate for being in the middle of a command pipeline (e.g. it should NOT pipe to another pager).

**Reply by PatrickF1** [reply]
This has been implemented in https://github.com/PatrickF1/fzf.fish/commit/90a66b91421f9d1459f5dec92d2a06d01164def3! Could someone mark this as the answer please?
